### PR TITLE
Cc shim stdin handling for terminal in raw mode

### PR DESCRIPTION
**Pull request by amshinde**

### CODE REVIEW

File: shim/shim.c [hunk 1]
```diff
@@ -270,38 +270,68 @@ handle_signals(char *container_id, int outfd) {
...
 {
-	ssize_t    nread;
-	char       buf[BUFSIZ-12];
-	char       wbuf[BUFSIZ];
-	int        ret;
-	ssize_t     len;
+	ssize_t      nread;
+	int          ret;
+	ssize_t      len;
+	static char  buf[BUFSIZ-12] = {0};
```
Comment (jodh-intel) [inline shim/shim.c]: What is `12`? `STREAM_HEADER_SIZE`?

File: shim/shim.c [hunk 2]
```diff
@@ -270,38 +270,68 @@ handle_signals(char *container_id, int outfd) {
...
+		}
+	} else if  (ch == '\r') {
+		emit = true;
+		cr = true;
+		buf[cnt++] = '\n';
+	} else {
+		cr = false;
+		buf[cnt++] = ch;
+	}
+	if (emit || cnt == (BUFSIZ-12)) {
```
Comment (jodh-intel) [inline shim/shim.c]: Can this magic number be replaced with a define?

**Comment by amshinde** [comment]
@jodh-intel Magic numbers are gone now.

**Comment by gorozco1** [comment]
👍

File: shim/shim.c [hunk 3]
```diff
@@ -270,38 +270,68 @@ handle_signals(char *container_id, int outfd) {
...
+	static int   cnt = 0;
+	char         ch;
 
 	if (! shim || shim->proxy_io_fd < 0) {
 		return;
 	}
 
-	// write data to I/O fd
-	while ((nread = read(STDIN_FILENO, buf, BUFSIZ-12)) != -1) {
+	nread = read(STDIN_FILENO , &ch, 1);
```
Comment (jodh-intel) [inline shim/shim.c]: Is this single-byte `read()` purely to simplify the handling of special characters?

**Comment by jodh-intel** [comment]
lgtm.

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
this was a false negative result : not ok 42 start and kill state
moving to passed

**Comment by chavafg** [comment]
qa-passed

**Pull request merged**